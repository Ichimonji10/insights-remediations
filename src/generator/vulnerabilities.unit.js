'use strict';

const { request, reqId } = require('../test');

test('generates a rule-based playbook', () => {
    const data = {
        issues: [{
            id: 'advisor:CVE_2017_6074_kernel|KERNEL_CVE_2017_6074',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459']
        }]
    };

    return request
    .post('/v1/playbook')
    .send(data)
    .expect(200)
    .then(res => expect(res.text).toMatchSnapshot());
});

test('generates a rule-based playbook with resolution preference', () => {
    const data = {
        issues: [{
            id: 'advisor:CVE_2017_6074_kernel|KERNEL_CVE_2017_6074',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459'],
            resolution: 'selinux_mitigate'
        }]
    };

    return request
    .post('/v1/playbook')
    .send(data)
    .expect(200)
    .then(res => expect(res.text).toMatchSnapshot());
});

test('400s on unknown resolution type', () => {
    const {id, header} = reqId();

    return request
    .post('/v1/playbook')
    .set(header)
    .send({
        issues: [{
            id: 'advisor:CVE_2017_6074_kernel|KERNEL_CVE_2017_6074',
            systems: ['68799a02-8be9-11e8-9eb6-529269fb1459'],
            resolution: 'non-existing-resolution'
        }]
    })
    .expect(400)
    .then(({ body }) => {
        body.errors.should.eql([{
            id,
            status: 400,
            code: 'UNKNOWN_RESOLUTION',
            title: 'Issue "advisor:CVE_2017_6074_kernel|KERNEL_CVE_2017_6074"' +
                ' does not have Ansible resolution "non-existing-resolution"'
        }]);
    });
});
