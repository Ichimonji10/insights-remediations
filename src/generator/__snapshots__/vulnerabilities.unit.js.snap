// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a rule-based playbook 1`] = `
"---
# Kernel vulnerable to local privilege escalation via DCCP module (CVE-2017-6074)
# Identifier: (vulnerabilities:CVE_2017_6074_kernel|KERNEL_CVE_2017_6074,kernel_update)
# Version: unknown
- name: Update system to the latest kernel and reboot
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: true
  vars:
    # determine if we need to update the 'kernel' package or 'kernel-rt' package
    kernel_pkg: \\"{{'kernel-rt' if 'rt' in ansible_kernel else 'kernel'}}\\"

  tasks:
    - name: Update kernel
      yum:
        name: \\"{{kernel_pkg}}\\"
        state: latest
      register: yum

    - when: yum|changed
      name: set reboot fact
      set_fact:
        insights_needs_reboot: True

    - when: not yum|changed
      # The latest kernel is already installed so boot from it.  Sort the installed kernels
      # by buildtime and select the one with the most recent build time
      block:
      - name: get latest installed {{kernel_pkg}} package version
        shell: rpm -q {{kernel_pkg}} --queryformat=\\"%{buildtime}\\\\t%{version}-%{release}.%{arch}\\\\n\\" | sort -nr | head -1 | cut -f2
        register: latest_kernel
        check_mode: no

      - name: get configured default kernel
        command: /sbin/grubby --default-kernel
        register: default_kernel
        check_mode: no

      - when: default_kernel.stdout.split('-', 1)[1] != latest_kernel.stdout
        name: set the default kernel to the latest installed
        command: /sbin/grubby --set-default /boot/vmlinuz-{{latest_kernel.stdout}}
        register: grub_change
        check_mode: no

      - when: grub_change|changed
        name: set reboot fact
        set_fact:
          insights_needs_reboot: True


# Reboots a system if any of the preceeding plays sets the 'insights_needs_reboot' variable to true.
# The variable can be overridden to suppress this behavior.
- name: Reboot system (if applicable)
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - when:
        - insights_needs_reboot is defined
        - insights_needs_reboot
      block:
        - name: Reboot system
          shell: sleep 2 && shutdown -r now \\"Ansible triggered reboot\\"
          async: 1
          poll: 0
          ignore_errors: true

        - name: Wait for system to boot up
          local_action:
            module: wait_for
            host: \\"{{ hostvars[inventory_hostname]['ansible_host'] | default(hostvars[inventory_hostname]['ansible_ssh_host'], true) | default(inventory_hostname, true) }}\\"
            port: \\"{{ hostvars[inventory_hostname]['ansible_port'] | default(hostvars[inventory_hostname]['ansible_ssh_port'], true) | default('22', true) }}\\"
            delay: 15
            search_regex: OpenSSH
            timeout: 300
          become: false

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: redhat-access-insights
      changed_when: false"
`;

exports[`generates a rule-based playbook with resolution preference 1`] = `
"---
- name: run insights to obtain latest report info
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  tasks:
    - name: determine insights version
      shell: 'redhat-access-insights --version'
      changed_when: false
      register: insights_version

    - when: insights_version.stdout[0:2] != '1.'
      block:
        - name: obtaining insights report
          shell: 'redhat-access-insights --to-json --quiet'
          register: insights_result
          changed_when: false
          check_mode: false
        - name: register insights report as fact for use by other plays
          set_fact: insights_report={{ insights_result.stdout }}

    - when: insights_version.stdout[0:2] == '1.'
      block:
        - name: obtaining insights report (legacy client)
          shell: 'redhat-access-insights --verbose | grep \\"Upload status: 201 Created\\" | grep -o \\"{.*}\\"'
          register: insights_result_legacy
          changed_when: false
          check_mode: false
        - name: register insights report as fact for use by other plays (legacy client)
          set_fact: insights_report={{ insights_result_legacy.stdout }}

# Kernel vulnerable to local privilege escalation via DCCP module (CVE-2017-6074)
# Identifier: (vulnerabilities:CVE_2017_6074_kernel|KERNEL_CVE_2017_6074,selinux_mitigate)
# Version: unknown
- name: Make sure SELinux is enabled, enforcing and has selinux-policy-3.13.1-81.el7 or later on RHEL7
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: yes
  vars:
    pydata: \\"{{insights_report.details['CVE_2017_6074_kernel|KERNEL_CVE_2017_6074']}}\\"

  tasks:
  - when: insights_report.details['CVE_2017_6074_kernel|KERNEL_CVE_2017_6074'] is defined
    block:
    - when: not pydata.selinux_can_help
      fail:
        msg: SELinux cannot mitigate the problem on this system.  Please try one of the other playbooks.

    - name: Get selinux mode
      command: getenforce
      register: getenforce
      check_mode: no

    - name: Enable SELinux immediately
      command: setenforce enforcing
      # setenforce enforcing will fail if SElinux is disabled so just ignore that situation
      failed_when: false

    - name: remove selinux=0 and enforcing=0 from grub config file
      command: /sbin/grubby --update-kernel=ALL --remove-args=\\"selinux enforcing\\"

    - name: Set SELINUX=enforcing in /etc/selinux/config
      lineinfile:
        backup: true
        dest: /etc/selinux/config
        regexp: '(?i)^\\\\s*SELINUX *=.*'
        line: SELINUX=enforcing
        state: present

    - when: ansible_distribution_major_version == '7' and
            pydata.minimal_selinux_policy and pydata.active_policy and
            pydata.minimal_selinux_policy | version_compare(pydata.active_policy, '>')
      name: Update selinux-policy package to latest version
      yum:
        name: selinux-policy
        state: latest

    - when: \\"getenforce.stdout == 'Disabled'\\"
      block:
      - name: SELinux relabel to be done on reboot (note, a relabel may take a while to complete)
        file:
          path: /.autorelabel
          state: touch

      - name: set reboot fact
        set_fact:
          insights_needs_reboot: True


# Reboots a system if any of the preceeding plays sets the 'insights_needs_reboot' variable to true.
# The variable can be overridden to suppress this behavior.
- name: Reboot system (if applicable)
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - when:
        - insights_needs_reboot is defined
        - insights_needs_reboot
      block:
        - name: Reboot system
          shell: sleep 2 && shutdown -r now \\"Ansible triggered reboot\\"
          async: 1
          poll: 0
          ignore_errors: true

        - name: Wait for system to boot up
          local_action:
            module: wait_for
            host: \\"{{ hostvars[inventory_hostname]['ansible_host'] | default(hostvars[inventory_hostname]['ansible_ssh_host'], true) | default(inventory_hostname, true) }}\\"
            port: \\"{{ hostvars[inventory_hostname]['ansible_port'] | default(hostvars[inventory_hostname]['ansible_ssh_port'], true) | default('22', true) }}\\"
            delay: 15
            search_regex: OpenSSH
            timeout: 300
          become: false

- name: run insights
  hosts: \\"68799a02-8be9-11e8-9eb6-529269fb1459.example.com\\"
  become: True
  gather_facts: False
  tasks:
    - name: run insights
      command: redhat-access-insights
      changed_when: false"
`;
